Ensure Rotation.to_euler() doesn't divide by zero to suppress warning

hakonanes opened the pull request:
Signed-off-by: Håkon Wiik Ånes <hwaanes@gmail.com>

#### Description of the change
* Ensure Rotation.to_euler() doesn't divide by zero to suppress warning
* Add test to ensure this

Closes #231.

#### Progress of the PR
- [n/a] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests with pytest for all lines
- [x] Clean code style by [running black via pre-commit](https://orix.readthedocs.io/en/latest/contributing.html#code-style)

#### For reviewers
<!-- Don't remove the checklist below. -->
- [x] The PR title is short, concise, and will make sense 1 year later.
- [n/a] New functions are imported in corresponding `__init__.py`.
- [n/a] New features, API changes, and deprecations are mentioned in the
      unreleased section in `CHANGELOG.rst`.

pc494 commented:
is this actually correct?

`np.divide(3,0)` 

return `inf` not zero so I am little lost.

harripj commented on orix/quaternion/rotation.py:
```diff
@@ -338,12 +338,15 @@ def to_euler(self, convention="bunge"):
             e[..., 2] = np.where(cond, 0, e[..., 2])
 
         if np.sum(chi != 0) > 0:
+            not_zero = chi != 0
```
If  `chi` contains floats (it looks like it is defined by a `sqrt` fn), then perhaps it would be better to use `~np.isclose(chi, 0)` here.

hakonanes commented:
> is this actually correct?

The warning is raised when the denominator is zero. We don't want inf, do we? Or do we want that? As in an array of inf should be passed to out?

@harripj I think the warning is given when chi is exactly zero, and that's the case I'd like to catch, not close to zero. Or do you think the warning is given when it is close to zero? If so, then your suggestion should be adopted.

harripj commented:
> @harripj I think the warning is given when chi is exactly zero, and that's the case I'd like to catch, not close to zero. Or do you think the warning is given when it is close to zero? If so, then your suggestion should be adopted.

It was my understanding that `islcose` should always be used with floating points, however at least on my machine `np.all(np.sqrt(np.zeros(50, dtype=float)) == 0)` is `True`, so it looks like the code above is safe.

The following also doesn't throw a warning: `np.divide(1, np.finfo(np.float64).min)`, so it seems to be good to go!

harripj commented:
I will let @pc494 pass an eye over this, but I am happy if this is merged.

hakonanes commented:
> It was my understanding that `islcose` should always be used with floating points rather than checking for exact equality, 

I didn't know this, do you know where you have it from? The documentation for [numpy.isclose()](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html) doesn't say anything about this.

Anyway, I changed the equality check to use `isclose()`.

> is this actually correct?
> np.divide(3,0)
> return inf not zero so I am little lost.

You're right, I've changed the output from `divide()` from 0 to `inf` when `chi` is zero.

Thank you both for good feedback!

pc494 commented:
I think we do want to passed array to be `inf`, otherwise we're not just removing a warning but actively changing the behaviour of the code (the `inf` gets turned into a finite number immediately by the `arctan2` so never makes it back to the user)

harripj commented:
> I didn't know this, do you know where you have it from? The documentation for numpy.isclose() doesn't say anything about this. Anyway, I changed the equality check to use isclose()

Sorry if I was confusing here, I think the previous logic was fine as the warning only arises for exact zeros and testing for exact zeros seems to be okay. 

> The following also doesn't throw a warning: np.divide(1, np.finfo(np.float64).min), so it seems to be good to go!

This should also have been written differently. The only case I can get force an error is division by precisely `0` or `1 / np.spacing(0)`, where `np.spacing(0) = 5e-324` on my machine.

hakonanes commented:
> Sorry if I was confusing here, I think the previous logic was fine as the warning only arises for exact zeros and testing for exact zeros seems to be okay.

You weren't, I decided to trust your instinct and go with a broader definition of when chi is zero.

pc494 commented:
> > It was my understanding that `islcose` should always be used with floating points rather than checking for exact equality,
> 
> I didn't know this, do you know where you have it from? The documentation for [numpy.isclose()](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html) doesn't say anything about this.
> 
> Anyway, I changed the equality check to use `isclose()`.
> 
> > is this actually correct?
> > np.divide(3,0)
> > return inf not zero so I am little lost.
> 
> You're right, I've changed the output from `divide()` from 0 to `inf` when `chi` is zero.
> 
> Thank you both for good feedback!

I'm merging now, but for what it's worth it's a shame the testing couldn't catch this...
